domain/application/state/containerimagemetadata.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/logger"
+	coreresource "github.com/juju/juju/core/resources"
+	"github.com/juju/juju/core/resources/containermetadataresource"
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/domain/application"
+	applicationerrors "github.com/juju/juju/domain/application/errors"
+	"github.com/juju/juju/internal/errors"
+)
+
+// ContainerImageMetadataState describes retrieval and persistence methods for storage.
+type ContainerImageMetadataState struct {
+	*domain.StateBase
+	logger logger.Logger
+}
+
+// NewContainerImageMetadataState returns a new state reference.
+func NewContainerImageMetadataState(factory database.TxnRunnerFactory, logger logger.Logger) *ContainerImageMetadataState {
+	return &ContainerImageMetadataState{
+		StateBase: domain.NewStateBase(factory),
+		logger:    logger,
+	}
+}
+
+// Get returns the container image metadata with the given UUID.
+// applicationerrors.ResourceNotFound is returned if the UUID is not in the
+// table
+func (s *ContainerImageMetadataState) Get(
+	ctx context.Context,
+	resourceUUID coreresource.ID,
+) (application.ContainerImageMetadata, error) {
+	db, err := s.DB()
+	if err != nil {
+		return application.ContainerImageMetadata{}, errors.Capture(err)
+	}
+
+	storageKey := containerImageMetadataStorageKey{StorageKey: resourceUUID.String()}
+	m := containerImageMetadata{}
+	stmt, err := s.Prepare(`
+SELECT &containerImageMetadata.*
+FROM   resource_oci_image_metadata_store
+WHERE  storage_key = $containerImageMetadataStorageKey.storage_key
+`, storageKey, m)
+	if err != nil {
+		return application.ContainerImageMetadata{}, errors.Errorf(
+			"preparing select OCI image resource metadata statement: %w", err)
```

Comment:
**suggestion:** subject of discussion, so don't take blindly this opinion: I don't think it's useful to decorate error in state level, especially prepare method.

Prepare method aren't expected to fail in production (those are always development issue). They would fail during development process, thus for me it is good enough to simply put `errors.Capture(err)`, with an ` gc.Commentf("something useful here: %v", errors.ErrorStack(err)` in TU.

Would be enough to quickly analyze and find the faulty prepare statement.

For Query error, my two cents would be: don't decorate if there is just one query. However, in case of several query, it can be useful.

Reply:
That's fair enough, though I find that annotating the prepares makes debugging my own code a lot easier. I don't have to worry about the error stack, I can just see from the message exactly where the error comes from. For errors like these, I lean more towards having enough context to make it clear from inspection of the error.

I agree with you that there's no point annotating single queries in the function, I'll remove that.

I like the idea of commenting the error stacks in the tests, I'll try to start doing that.